### ISSUE TITLE
Third party keyboards don't play nicely with layout

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#4029]
## Steps to Reproduce

1) Install GBoard for iOS
2) Run a Flutter app with an Input widget (I used the codelab app)
## Expected behavior

The app should layout in response to the existence of the keyboard.
## Actual behavior

With the native iOS keyboard, you get the correct behavior, but with the extension keyboard (GBoard), the app doesn't seem to know the keyboard is active.

![keyboard-native](https://cloud.githubusercontent.com/assets/15253456/15378865/2729ee68-1d1c-11e6-959a-1424f4f55a47.png)

![keyboard-gboard](https://cloud.githubusercontent.com/assets/15253456/15378869/2c2c1062-1d1c-11e6-83c6-c11f5cd62a1f.png)
## Flutter Doctor

N/A
## Logs and Crash Reports

Nothing in the logs

**Comment by abarth** [comment]
Do we not get the same notification from UIKit?

**Comment by tvolkert** [comment]
Not sure - I'll investigate

**Comment by tvolkert** [comment]
Incidentally, I noticed that the native Drive app suffers from this as well (cursor is hidden underneath GBoard when the keyboard opens, and the user has to scroll down to reveal the cursor again).  It was exacerbated in the codelab because the basic layout wasn't scrollable.  I'll see if there are any notifications from UIKit we could pick up on, but if the answer's no, then the user will just have to scroll...

**Comment by abarth** [comment]
We currently listen for UIKeyboardDidShowNotification.  We get the size of the keyboard from:

```
  CGFloat bottom = CGRectGetHeight(
      [[info objectForKey:UIKeyboardFrameBeginUserInfoKey] CGRectValue]);
```

Maybe there's a better way?

**Comment by eseidelGoogle** [comment]
This may be fixed by @cbracken's keyboard work.

**Comment by eseidelGoogle** [comment]
See also https://github.com/flutter/flutter/issues/8777.

**Comment by eseidelGoogle** [comment]
@cbracken may have already fixed this.  Otherwise we can punt to M3.

**Comment by cbracken** [comment]
My work doesn't fix this, but it's very nearby and should be a pretty easy fix. If no one else gets to it first, I'm happy to take a look once the higher-priority text-related bugs are cleaned out.

**Comment by cbracken** [comment]
Fixed by flutter/engine#3681 and rolled in #9978
